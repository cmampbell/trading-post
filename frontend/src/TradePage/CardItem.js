import React from 'react';
import Card from '@mui/material/Card';
import Typography from '@mui/material/Typography';
import Grid from '@mui/material/Grid';
import Button from '@mui/material/Button';
import CloseIcon from '@mui/icons-material/Close';
import PriceDisplay from '../_common/PriceDisplay';

/* Component used to display card info on the trade page.
*
*  Props:
*       card - the card to display
*       deleteCard - function to remove the card from trade page state
*/

const CardItem = ({ card, deleteCard }) => {
    const {
        name,
        set_code: setCode,
        collector_number: collectorNumber,
        foil,
        quantity,
        id: cardId
    } = card;

    const deleteSelf = () => {
        deleteCard(cardId)
    }

    return (
        <Card sx={
            foil === 'Yes' ? {
                border: '2px solid transparent',
                borderImage: 'linear-gradient(to bottom right, #b827fc 0%, #2c90fc 25%, #b8fd33 50%, #fec837 75%, #fd1892 100%)',
                borderImageSlice: 1,
            }
                : {
                    border: 2,
                    borderColor: 'black',
                    width: 'auto',
                }}>
            <Grid container spacing={0} sx={{ width: '95%' }}>
                <Grid item xs={8}>
                    <Typography variant='body1' sx={{ textAlign: 'left', fontWeight: 'bold', paddingLeft: '3vw' }}> {name} </Typography>
                </Grid>
                <Grid item xs={3}>
                    <Typography variant='body1'>
                        Qty: {quantity}
                    </Typography>
                </Grid>
                <Grid item xs={1}>
                    <Button onClick={deleteSelf} className={name} color='error' sx={{ top: '-13%' }} >
                        <CloseIcon />
                    </Button>
                </Grid>
                <Grid item xs={6}>
                    <Typography variant='body2' sx={{ textAlign: 'left', paddingLeft: '3vw' }}>
                        {setCode.toUpperCase()} - #{collectorNumber}
                    </Typography>
                </Grid>
                <Grid item xs={2}>

                </Grid>
                <Grid item xs={4}>
                    <PriceDisplay card={card} sx={{ textAlign: 'left', paddingLeft: '4.5vw' }} />
                </Grid>
            </Grid>
        </Card >
    )
}

export default CardItem;