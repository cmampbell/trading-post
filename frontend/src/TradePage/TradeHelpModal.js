import React from 'react';
import Modal from "@mui/material/Modal";
import Container from "@mui/material/Container";
import Paper from "@mui/material/Paper";
import Typography from "@mui/material/Typography";
import List from "@mui/material/List";
import ListItem from "@mui/material/ListItem";

/* Returns MUI modal component https://mui.com/material-ui/react-modal/
*
*  Displays text instructions on how to use the Trade Page.
*
*  Props:
*       open - boolean used to track status of modal.
*       toggleHelpModal - function used to open and close Modal by toggling open boolean
*/

const TradeHelpModal = ({ open, toggleHelpModal }) => {

    const handleClose = () => {
        toggleHelpModal(false)
    }

    return (
        <Modal
            open={open}
            onClose={handleClose}
            aria-labelledby="modal-title"
            aria-describedby="modal-description"
        >
            <Container maxWidth="md">
                <Paper elevation={8}
                    sx={{
                        position: 'relative',
                        display: 'flex',
                        justifyContent: 'flex-start',
                        flexDirection: 'column',
                        alignItems: 'center',
                        marginTop: '8vh',
                        height: 'auto'
                    }}
                >
                    <List>
                        <ListItem>
                            <Typography variant="body1">
                                {`1)`} Click a card list tab make it active
                            </Typography>
                        </ListItem>
                        <ListItem>
                            <Typography variant="body1">
                                {`2)`} Click the 'Add Card' button to add a card to the list
                            </Typography>
                        </ListItem>
                        <ListItem>
                            <Typography variant="body1">
                                {`3)`} Using the search bar or camera button, add your cards to each list.
                            </Typography>
                        </ListItem>
                        <ListItem>
                            <Typography variant="body1">
                                {`4)`} When both lists are complete, compare the values at the bottom price and decide whether to make the trade!
                            </Typography>
                        </ListItem>
                    </List>
                </Paper>
            </Container>
        </Modal >
    )
}

export default TradeHelpModal;