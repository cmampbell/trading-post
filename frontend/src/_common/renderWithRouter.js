import React, { isValidElement } from "react";
import { render } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router-dom";


// borrowed from: https://webup.org/blog/how-to-avoid-mocking-in-react-router-v6-tests/

export default renderWithRouter = (children, routes=[]) => {
    // options creating routes for valid children
    // else keep children as they are
    // first argument is a full route

    const options = isValidElement(children)
            ? { element: children, path: '/'}
            : children;

    // spread out the options here
    // the options are a route object
    // then routes are spread and passed after
    const router = createMemoryRouter([{...options}, ...routes],
        {
        initialEntries: [options.path],
        initialIndex: 1
    }
    );

    return render(<RouterProvider router={router}/>);
}