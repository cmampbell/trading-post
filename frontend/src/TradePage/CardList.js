import React, { useState } from 'react';
import { Button, Stack } from "@mui/material";
import CardItem from './CardItem';
import AddCardModal from '../_common/AddCardModal/AddCardModal';
import SetSelectField from "../_common/CardForm/FormInputs/SetSelectField";
import FoilSelectField from "../_common/CardForm/FormInputs/FoilSelectField";
import QualitySelectField from "../_common/CardForm/FormInputs/QualitySelectField";
import QuantitySelectField from "../_common/CardForm/FormInputs/QuantitySelectField";

const CardList = ({ cards = [], addToCardLists, removeFromCardLists }) => {
    const [searchOpen, setSearchOpen] = useState(false);

    const handleSearchOpen = () => {
        setSearchOpen(true);
    }

    return (
        <>
        <Button onClick={handleSearchOpen} variant="outlined" sx={{marginTop: '1vh'}} color='secondary'>Add Card</Button>
            <Stack spacing={1} sx={{marginTop: '1vh', marginBottom: '3vh'}}>
            
                {cards.map((card, idx) => <CardItem card={card} key={`${card.id}+${idx}`} deleteCard={removeFromCardLists} />)}
            </Stack>
             
            <AddCardModal open={searchOpen} setSearchOpen={setSearchOpen} addCard={addToCardLists} fields={[SetSelectField, FoilSelectField, QualitySelectField, QuantitySelectField]} />
        </>
    )
}

export default CardList;