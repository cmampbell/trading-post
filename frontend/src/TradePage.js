import React, { useState } from "react";
import { Grid } from "@mui/material"

import CardList from './CardList'

const TradePage = () => {
    // Render two card list components
    // we can handle grid and formatting here

    // Render two CardList components
    // handle which list is active here

    const [activeList, setActiveList] = useState('left')

    const makeActive = (side) => {
        setActiveList(side)
    }

    return (
        <Grid container spacing={2}>
            <Grid item xs={12}>
                <h1>Trade Details</h1>
            </Grid>
            <CardList activeList={activeList} makeActive={makeActive} side='left'/>
            <CardList activeList={activeList} makeActive={makeActive} side='right'/>
        </Grid>
    )

    // Trade page title
    // 2 Card List Components
    // 1 left - current user will always be on right
    // left will always start active
    // 1 right - other user will be on the left
}

export default TradePage;