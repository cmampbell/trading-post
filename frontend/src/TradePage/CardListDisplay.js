import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import Box from '@mui/material/Box';
import CardList from './CardList';
import { Container, Typography } from '@mui/material';
import { styled } from '@mui/material/styles';
import PriceDisplay from '../_common/PriceDisplay';

/* Returns a MUI Container with Tabs for each Card List
*
*  Uses code from here for Tab functionality: https://mui.com/material-ui/react-tabs/
*
*  State: value - the index prop of the current tab that is open
*
*         cardLists - an object containing keys of tab indexes, with values
*                     of card arrays. We track the state of each list here.
*
*/

function CustomTabPanel(props) {
    const { children, value, index, ...other } = props;

    return (
        <div
            role="tabpanel"
            hidden={value !== index}
            id={`simple-tabpanel-${index}`}
            aria-labelledby={`simple-tab-${index}`}
            {...other}
        >
            {value === index && (
                <Box sx={{ borderLeft: 1, borderRight: 1, borderBottom: 1, borderColor: 'divider' }}>
                    {children}
                </Box>
            )}
        </div>
    );
}

CustomTabPanel.propTypes = {
    children: PropTypes.node,
    index: PropTypes.number.isRequired,
    value: PropTypes.number.isRequired,
};

function a11yProps(index) {
    return {
        id: `card-list-tab-${index}`,
        'aria-controls': `card-list-tabpanel-${index}`,
    };
}

const StyledTabs = styled((props) => (
    <Tabs
        {...props}
        TabIndicatorProps={{ children: <span className="MuiTabs-indicatorSpan" /> }}
    />
))({
    '& .MuiTabs-indicator': {
        display: 'flex',
        justifyContent: 'center',
        backgroundColor: 'transparent',
    },
    '& .MuiTabs-indicatorSpan': {
        maxWidth: 40,
        width: '100%',
        backgroundColor: '#01589b',
    },
});

export default function CardListDisplay() {
    const [value, setValue] = useState(0);
    const [cardLists, setCardLists] = useState({ 0: [], 1: [] })

    const addToCardLists = (card, cardData) => {
        setCardLists((oldCardLists) => ({ ...oldCardLists, [value]: [...oldCardLists[value], {...card, ...cardData}] }))
    }

    const removeFromCardLists = (cardId) => {
        setCardLists((oldCardLists) => ({ ...oldCardLists, [value]: oldCardLists[value].filter((card) => card.id !== cardId) }));
    }

    const handleChange = (event, newValue) => {
        setValue(newValue);
    };

    return (
        <Container sx={{ width: '95vw' }}>
            <StyledTabs
                value={value}
                onChange={handleChange}
                aria-label="card list tabs"
                variant="fullWidth"
            >
                <Tab
                    label="Card List One"
                    {...a11yProps(0)}
                    sx={{ borderTop: 2, borderRight: 2, borderLeft: 2, borderBottom: value === 0 ? 0 : 2, borderColor: 'divider' }}
                />
                <Tab
                    label="Card List Two"
                    {...a11yProps(1)}
                    sx={{ borderTop: 2, borderRight: 2, borderBottom: value === 1 ? 0 : 2, borderColor: 'divider' }}
                />
            </StyledTabs>
            <CustomTabPanel value={value} index={0}>
                <CardList
                    value={value}
                    addToCardLists={addToCardLists}
                    removeFromCardLists={removeFromCardLists}
                    cards={cardLists[0]}
                />
            </CustomTabPanel>
            <CustomTabPanel value={value} index={1}>
                <CardList
                    value={value}
                    addToCardLists={addToCardLists}
                    removeFromCardLists={removeFromCardLists}
                    cards={cardLists[1]}
                />
            </CustomTabPanel>
            <Typography>
                Card List One Price: <PriceDisplay cards={cardLists[0]} />
            </Typography>
            <Typography>
                Card List Two Price: <PriceDisplay cards={cardLists[1]} />
            </Typography>
        </Container>
    );
}