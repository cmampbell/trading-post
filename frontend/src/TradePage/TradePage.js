import React, { useState } from "react";
import Grid from "@mui/material/Grid";
import Typography from "@mui/material/Typography";
import IconButton from "@mui/material/IconButton";
import Box from "@mui/material/Box";
import TradeHelpModal from "./TradeHelpModal";
import HelpIcon from '@mui/icons-material/Help';
import CardListDisplay from "./CardListDisplay";

/* Parent component for trade page. 
*  Renders page title, <TradeHelpModal/>, and <CardListDisplay/>.
*
*  State:
*       - showHelp - boolean used to track <TradeHelpModal/> open status
*/
const TradePage = () => {
    const [showHelp, setShowHelp] = useState(false);

    const toggleHelpModal = () => {
        setShowHelp((oldShowHelp) => !oldShowHelp);
    };

    return (
        <Grid container spacing={2} sx={{ position: 'relative' }}>
            <Grid item xs={12}>
                <Box sx={{ position: 'relative' }}>
                    <Typography variant="h1">
                        Trade Page
                    </Typography>
                    <IconButton
                        onClick={toggleHelpModal}
                        sx={{
                            position: "absolute",
                            top: '25%',
                            right: '1%'
                        }}
                        color="info"
                    >
                        <HelpIcon />
                    </IconButton>
                </Box>
                <TradeHelpModal open={showHelp} toggleHelpModal={toggleHelpModal} />
            </Grid>
            <CardListDisplay />
        </Grid>
    );
};

export default TradePage;