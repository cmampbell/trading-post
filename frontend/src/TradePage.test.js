import React from "react";
import renderWithRouter from "./renderWithRouter";
import TradePage from "./TradePage";
import userEvent from '@testing-library/user-event'
import { act } from "react-dom/test-utils";
import { waitFor, screen } from "@testing-library/react";

const user = userEvent.setup()

describe('HomePage tests', () => {
    it('should render without crashing', () => {
        // define a full route as the first argument
        renderWithRouter(<TradePage />);
    })

    it('should match the snapshot', () => {
        const { asFragment } = renderWithRouter(<TradePage />);

        expect(asFragment()).toMatchSnapshot();
    })

    it('should have correct elements', () => {
        const { queryAllByRole, queryByText } = renderWithRouter(<TradePage />);

        expect(queryByText('Trade Page')).toBeInTheDocument();
        expect(queryAllByRole("list").length).toEqual(2);
        expect(queryByText('Commit Trade')).toBeInTheDocument();
    })
})