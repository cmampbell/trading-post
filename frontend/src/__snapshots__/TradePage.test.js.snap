// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage tests should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    TradePage
  </h1>
</DocumentFragment>
`;
